### PR TITLE
fix #287120, fix #282503: incomplete filling of system

### PR DESCRIPTION
**Pull request by MarcSabatella**
See analysis in https://musescore.org/en/node/287120, also report in https://musescore.org/en/node/282503.  In our partial relayout scheme, we were considering the job done as soon as we processed the range actually containing the change and reached the end of that system.  But this is just wrong; it could easily be the case that the change results in more measure possibly fitting on the system.  And there was a TODO in the code for this.

This PR implements what we should have been doing all along: not stopping as soon as we reach the end of the current system, but continuing collecting measures to fill the system, and not stopping until we actually reach a system whose final measure is the same as on the previous layout.  To make this work, we also need to preserve some state of the next measure, because in the process of collecting measures, we keep mucking with the header, trailer, and width, so by the time we realize we are done, we've actually processed one measure too many and need to restore it.

I've tested this on all of those cases and haven't made it fail.  Worst case if I've missed something is probably something akin to what happens now, some aspect of layout of the next system is off until the next full relayout.

### CODE REVIEW

File: libmscore/layout.cpp [hunk 1]
```diff
@@ -3287,6 +3287,11 @@ System* Score::collectSystem(LayoutContext& lc)
       qreal systemWidth = styleD(Sid::pagePrintableWidth) * DPI;
       system->setWidth(systemWidth);
 
+      // save state of measure
+      qreal curWidth = lc.curMeasure->width();;
+      bool curHeader = lc.curMeasure->header();;
```
Comment (dmitrio95) [inline libmscore/layout.cpp]: Extra semicolons

File: libmscore/layout.cpp [hunk 2]
```diff
@@ -3402,28 +3407,55 @@ System* Score::collectSystem(LayoutContext& lc)
                         break;
                   }
 
+            // preserve state of next measure (which is about to become current measure)
+            if (lc.nextMeasure) {
+                  curWidth = lc.nextMeasure->width();;
+                  curHeader = lc.nextMeasure->header();;
```
Comment (dmitrio95) [inline libmscore/layout.cpp]: Extra semicolons too

**Comment by MarcSabatella** [comment]
I like the idea of refactoring, but there are some complications here.  Currently, we *don't* use exactly the same code to process the measure inside and outside the loop, because inside the loop we need to do different things depending on whether it's the first measure or not.  On the other hand, that in itself leads to problems, I just realized this is probably we sometimes get issues in systems with only a single measure.  So I will look at how to best clean this up.  And of course, remove the extra semicolons!

**Comment by dmitrio95** [comment]
@MarcSabatella will you do an update to this PR or is it better to leave the refactoring for a separate patch?

**Comment by MarcSabatella** [comment]
Weird, I thought I had already updated for the semicolons at least.  And I know I did look at what made sense refactoring-wise.  Will look again.

**Comment by MarcSabatella** [comment]
I *had* removed the extra semicolons locally but never pushed it.  Done now.

I also looked at what made sense to refactor, but didn't come up with anything.  As I above, we're not really doing the same thing, for good reason.  I did try for a while to see if I could reproduce any of the problems I remembered seeing in the past involving courtesy elements on single line systems, but I did not see any issues (possibly this PR itself fixes some).

Given that I don't see a way to refactor without actually changing some aspect of behavior, and things are working well as is with this PR, I'd rather not risk trying to force a refactor unnecessarily.  So, bottom line, if there is to be refactoring, it's another job for another time.

**Comment by dmitrio95** [comment]
OK, let's leave a possible refactoring for a later time then.

**Pull request merged**